ctapipe/reco/sklearn.py

```diff
@@ -0,0 +1,590 @@
+
+
+class EnergyRegressor(SKLearnRegressionReconstructor):
+    """
+    Use a scikit-learn regression model per telescope type to predict primary energy
+    """
+
+    #: Name of the target table column for training
+    target = "true_energy"
+    property = "energy"
+
+    def __call__(self, event: ArrayEventContainer) -> None:
+        """
+        Apply model for a single event and fill result into the event container
+        """
+        for tel_id in event.trigger.tels_with_trigger:
+            table = collect_features(event, tel_id, self.instrument_table)
+            table = self.generate_features(table)
+
+            passes_quality_checks = self.qualityquery.get_table_mask(table)[0]
+
+            if passes_quality_checks:
+                prediction, valid = self._predict(
+                    self.subarray.tel[tel_id],
+                    table,
+                )
+                container = ReconstructedEnergyContainer(
+                    energy=prediction[0],
+                    is_valid=valid[0],
+                )
+            else:
+                container = ReconstructedEnergyContainer(
+                    energy=u.Quantity(np.nan, self.unit),
+                    is_valid=False,
+                )
+
+            container.prefix = f"{self.model_cls}_tel"
+            event.dl2.tel[tel_id].energy[self.model_cls] = container
+
+        self.stereo_combiner(event)
+
+    def predict_table(self, key, table: Table) -> Table:
+        """Predict on a table of events"""
+        table = self.generate_features(table)
+
+        n_rows = len(table)
+        energy = u.Quantity(np.full(n_rows, np.nan), self.unit, copy=False)
+        is_valid = np.full(n_rows, False)
+
+        valid = self.qualityquery.get_table_mask(table)
+        energy[valid], is_valid[valid] = self._predict(key, table[valid])
+
+        result = Table(
+            {
+                f"{self.model_cls}_tel_energy": energy,
+                f"{self.model_cls}_tel_is_valid": is_valid,
+            }
+        )
+        add_defaults_and_meta(
+            result,
+            ReconstructedEnergyContainer,
+            prefix=self.model_cls,
+            stereo=False,
+        )
+        return result
+
+
+class ParticleClassifier(SKLearnClassficationReconstructor):
+    """
+    Predict dl2 particle classification
+    """
+
+    #: Name of the target table column for training
+    target = "true_shower_primary_id"
+    property = "classification"
+
+    positive_class = Integer(
+        default_value=0,
+        help="Particle id (in simtel system) of the positive class. Default is 0 for gammas.",
+    ).tag(config=True)
+
+    def __call__(self, event: ArrayEventContainer) -> None:
+        for tel_id in event.trigger.tels_with_trigger:
+            table = collect_features(event, tel_id, self.instrument_table)
+            table = self.generate_features(table)
+            mask = self.qualityquery.get_table_mask(table)
```

Comment:
This is inconsistent with above, where the 0th value is taken directly and the result named `passes_quality_checks` (l366)